storage/generic.go

```diff
@@ -0,0 +1,88 @@
+package storage
+
+import "github.com/prometheus/prometheus/pkg/labels"
+
+// Boilerplate on purpose. Generics some day...
+
+type genericQuerierAdapter struct {
+	baseQuerier
+
+	// Union.
+	q  Querier
+	cq ChunkedQuerier
+}
+
+func (q *genericQuerierAdapter) Select(params *SelectParams, matchers ...*labels.Matcher) (genericSeriesSet, Warnings, error) {
+	if q.q != nil {
+		s, w, err := q.q.Select(params, matchers...)
+		return s.(genericSeriesSet), w, err
+	}
+	s, w, err := q.cq.Select(params, matchers...)
+	return s.(genericSeriesSet), w, err
+}
+func (q *genericQuerierAdapter) SelectSorted(params *SelectParams, matchers ...*labels.Matcher) (genericSeriesSet, Warnings, error) {
+	if q.q != nil {
+		s, w, err := q.q.SelectSorted(params, matchers...)
+		return s.(genericSeriesSet), w, err
+	}
+	s, w, err := q.cq.SelectSorted(params, matchers...)
+	return s.(genericSeriesSet), w, err
+}
+
+func newGenericQuerierFrom(q Querier) genericQuerier {
+	return &genericQuerierAdapter{baseQuerier: q, q: q}
+}
+
+func newGenericQuerierFromChunked(cq ChunkedQuerier) genericQuerier {
+	return &genericQuerierAdapter{baseQuerier: cq, cq: cq}
+}
+
+type genericQuerier interface {
+	baseQuerier
+	Select(*SelectParams, ...*labels.Matcher) (genericSeriesSet, Warnings, error)
+	SelectSorted(*SelectParams, ...*labels.Matcher) (genericSeriesSet, Warnings, error)
+}
+
+type genericSeriesSet interface {
+	Next() bool
+	At() Labeled
```

Comment:
```suggestion
	At() Series
```

Reply:
Nope, we need Labelled, but also because it's actually `genericSeries` ;p